### PR TITLE
chore: docker helper 4x move curl TLS certs outside nginx container

### PR DESCRIPTION
**Pull request by mrjones-plip**
# Description

In the forums folks have [had issues](https://forum.communityhealthtoolkit.org/t/using-sudo-with-docker-helper-installing-tls-certs/3979/9?u=mrjones) getting TLS certs downloaded in docker helper.  This PR:
* updates [add cert script](https://github.com/medic/cht-core/blob/master/scripts/add-local-ip-certs-to-docker-4.x.sh) and [docker helper](https://github.com/medic/cht-core/blob/master/scripts/docker-helper-4.x/cht-docker-compose.sh) to move the `curl` calls to be on `localhost` instead of inside `nginx` container.
* be extra defensive about `curl` calls to output user facing errors in case of DNS or `404` errors
* add an explicit check for valid TLS after we install the cert
* unchanged: we re-install the cert every time clean
* updates docs URLs to not be redirects

## Testing
* test on a new install of CHT `4.10` in docker helper - ensure valid cert
* on the instance you just created, use the [add cert script](https://github.com/medic/cht-core/blob/master/scripts/add-local-ip-certs-to-docker-4.x.sh) to install a self signed cert and re-install  - ensure valid cert
* on an existing docker helper instance created not in this branch - re-run docker helper  - ensure valid cert

## Failure mode

This is is what it looks like when TLS installation fails.  This was achieved by [changing the local-ip cert](https://github.com/medic/cht-core/blob/57f8ed2b7672f334fbf125b015570de4d4b489a8/scripts/docker-helper-4.x/cht-docker-compose.sh#L274) to be a `404` URL

```bash
./cht-docker-compose.sh cleaaaaaaannnnn.env

homedir: /home/mrjones/.medic/cht-docker/cleaaaaaaannnnn-dir
WARN[0000] /home/mrjones/.medic/cht-docker/cleaaaaaaannnnn-dir/upgrade-service.yml: the attribute `version` is obsolete, it will be ignored, please remove it to avoid potential confusion 
[+] Running 1/0
 ✔ Container cleaaaaaaannnnn-dir-cht-upgrade-service-1  Running                                                                                                              0.0s 
Starting project "cleaaaaaaannnnn". First run takes a while. Will try for up to five minutes...
curl: (22) The requested URL returned error: 404

Failed to install local-ip TLS certificate. Check for errors above and try again

---DEBUG INFO---
Load: 1.16 1.09 1.37
CHT Containers: 7
Global Containers: 7
URL: https://192-168-68-26.local-ip.medicmobile.org:10451

Service              Status   Container                                  Image
cht-upgrade-service  running  cleaaaaaaannnnn-dir-cht-upgrade-service-1  public.ecr.aws/s5s3h4s7/cht-upgrade-service:latest
haproxy              running  cleaaaaaaannnnn-haproxy-1                  public.ecr.aws/medic/cht-haproxy:4.10.0
healthcheck          running  cleaaaaaaannnnn-healthcheck-1              public.ecr.aws/medic/cht-haproxy-healthcheck:4.10.0
api                  running  cleaaaaaaannnnn-api-1                      public.ecr.aws/medic/cht-api:4.10.0
sentinel             running  cleaaaaaaannnnn-sentinel-1                 public.ecr.aws/medic/cht-sentinel:4.10.0
nginx                running  cleaaaaaaannnnn-nginx-1                    public.ecr.aws/medic/cht-nginx:4.10.0
couchdb              running  cleaaaaaaannnnn-couchdb-1                  public.ecr.aws/medic/cht-couchdb:4.10.0
```

## Success mode

this is unchanged from before this PR and looks like:

```bash
./cht-docker-compose.sh cleaaaaaaannnnn.env

homedir: /home/mrjones/.medic/cht-docker/cleaaaaaaannnnn-dir
WARN[0000] /home/mrjones/.medic/cht-docker/cleaaaaaaannnnn-dir/upgrade-service.yml: the attribute `version` is obsolete, it will be ignored, please remove it to avoid potential confusion 
[+] Running 1/0
 ✔ Container cleaaaaaaannnnn-dir-cht-upgrade-service-1  Running                                                                                                              0.0s 
Starting project "cleaaaaaaannnnn". First run takes a while. Will try for up to five minutes...


 -------------------------------------------------------- 

  Success! "cleaaaaaaannnnn" is set up:

    https://192-168-68-26.local-ip.medicmobile.org:10451/ (CHT)
    https://192-168-68-26.local-ip.medicmobile.org:10451/_utils/ (Fauxton)

    Login: medic
    Password: password

 -------------------------------------------------------- 

Start existing project
    ./cht-docker-compose.sh ENV-FILE.env

Stop and keep project:
    ./cht-docker-compose.sh ENV-FILE.env stop

Stop and destroy all project data:
    ./cht-docker-compose.sh ENV-FILE.env destroy

https://docs.communityhealthtoolkit.org/hosting/4.x/app-developer/


 Have a great day! 
```

# Code review checklist

- [X] Readable: Concise, well named, follows the [style guide](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/), documented if necessary.
- [X] Tested: Unit and/or e2e where appropriate
- [X] Backwards compatible: Works with existing data and configuration or includes a migration. Any breaking changes documented in the release notes.

# Compose URLs
If Build CI hasn't passed, these may 404:

* [Core](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:docker-helper-curl-tls-move/docker-compose/cht-core.yml)
* [CouchDB Single](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:docker-helper-curl-tls-move/docker-compose/cht-couchdb.yml)
* [CouchDB Cluster](https://staging.dev.medicmobile.org/_couch/builds_4/medic:medic:docker-helper-curl-tls-move/docker-compose/cht-couchdb-clustered.yml)

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

### CODE REVIEW

File: scripts/docker-helper-4.x/cht-docker-compose.sh [hunk 1]
```diff
@@ -101,7 +101,7 @@ show_help_existing_stop_and_destroy() {
     echo "Stop and destroy all project data:"
     echo "    ./cht-docker-compose.sh ENV-FILE.env destroy"
     echo ""
-    echo "https://docs.communityhealthtoolkit.org/apps/guides/hosting/4.x/app-developer/"
+    echo "https://docs.communityhealthtoolkit.org/hosting/4.x/app-developer/"
```
Comment (m5r) [inline scripts/docker-helper-4.x/cht-docker-compose.sh:104]: Good catch :eyes:

File: scripts/docker-helper-4.x/cht-docker-compose.sh [hunk 2]
```diff
@@ -261,23 +262,45 @@ get_system_and_docker_info(){
...
+  docker cp /tmp/local-ip-key "${nginxContainerId}":/etc/nginx/private/key.pem  2>/dev/null
+  docker exec "$nginxContainerId" bash -c "nginx -s reload"  2>/dev/null
+}
+
+validate_tls(){
+  url=$1
+  # by default curl validates TLS.  If we get back  60 then TLS isn't valid:
+  # exitcode: https://everything.curl.dev/usingcurl/verbose/writeout.html
+  # 60: https://everything.curl.dev/cmdline/exitcode.html
+  status=$(curl  --write-out "%{exitcode}"  -qs  "$url" -o /dev/null)
```
Comment (m5r) [inline scripts/docker-helper-4.x/cht-docker-compose.sh:286]: Thanks for the command explanation

File: scripts/add-local-ip-certs-to-docker-4.x.sh [hunk 3]
```diff
@@ -36,6 +36,14 @@
 # ./add-local-ip-certs-to-docker-4.x.sh cht-3-14 expire
 #
 
+update_nginx_local_ip_tls_cert(){
+  nginxContainerId=$1
+  curl --fail --silent --show-error -o /tmp/local-ip-fullchain https://local-ip.medicmobile.org/fullchain
+  curl --fail --silent --show-error -o /tmp/local-ip-key https://local-ip.medicmobile.org/key
```
Comment (m5r) [inline scripts/add-local-ip-certs-to-docker-4.x.sh:42]: What do you think of making `curl` retry the requests to cover for the occasional networking issue?

https://stackoverflow.com/questions/42873285/curl-retry-mechanism/42873372#42873372

```suggestion
  curl --retry 3 --fail --silent --show-error -o /tmp/local-ip-fullchain https://local-ip.medicmobile.org/fullchain
  curl --retry 3 --fail --silent --show-error -o /tmp/local-ip-key https://local-ip.medicmobile.org/key
```

**Comment by m5r** [comment]
I ran into this error message `Failed to install local-ip TLS certificate. Check for errors above and try again` without any other error printed out.

image

Checking with both `curl` and in a browser, the certificate is valid :raised_eyebrow: 
`curl --write-out "%{exitcode}" -qs https://192-168-1-29.local-ip.medicmobile.org:10443 -o /dev/null` returns `0` and:
image

File: scripts/docker-helper-4.x/cht-docker-compose.sh (hunk 2 above)
Comment (mrjones-plip) [reply scripts/docker-helper-4.x/cht-docker-compose.sh:286]: curl is powerful, but can be a bit obtuse!

File: scripts/add-local-ip-certs-to-docker-4.x.sh (hunk 3 above)
Comment (mrjones-plip) [reply scripts/add-local-ip-certs-to-docker-4.x.sh:42]: wow - excellent thinking!  Is there anything curl can't do?!?! [TIL](https://everything.curl.dev/usingcurl/downloads/retry.html)!

**Comment by mrjones-plip** [comment]
> I'm thinking there might be a race condition where the script runs the TLS validation while the nginx container isn't up yet

ah - yeah - this could be!  Let's add a ....retry!  I'll also want to test with the retries that added in 57d839f  anyway , so lemme do some local work here.

**Comment by mrjones-plip** [comment]
local testing showed we could show `stderr` and ignore `stdout` on `docker cp` commands which help highlight failure scenarios. 

tested `retry` in `curl` calls which hopefully addresses too quickly testing TLS before `nginx` has had a chance to restart. 

see a698e5d0d

File: scripts/docker-helper-4.x/cht-docker-compose.sh [hunk 4]
```diff
@@ -261,23 +262,45 @@ get_system_and_docker_info(){
...
+  docker cp /tmp/local-ip-key "${nginxContainerId}":/etc/nginx/private/key.pem  1>/dev/null
+  docker exec "$nginxContainerId" bash -c "nginx -s reload"  2>/dev/null
+}
+
+validate_tls(){
+  url=$1
+  # by default curl validates TLS.  If we get back  60 then TLS isn't valid:
+  # exitcode: https://everything.curl.dev/usingcurl/verbose/writeout.html
+  # 60: https://everything.curl.dev/cmdline/exitcode.html
+  status=$(curl   --retry 3 --write-out "%{exitcode}"  -qs  "$url" -o /dev/null)
```
Comment (m5r) [inline scripts/docker-helper-4.x/cht-docker-compose.sh:286]: It looks like `--retry` doesn't actually retry in some cases :disappointed: I'm still running into the `Failed to install...` error

While spinning up a CHT instance with docker helper, I ran this command over and over in a different terminal:
```sh
curl --retry 3 --write-out "%{exitcode} %{num_retries}" --fail -qs https://192-168-1-29.local-ip.medicmobile.org:10443 -o /dev/null
```
It prints the exit code and the number of retries performed. For the exit codes 7 and 60 (respectively "Failed to connect to host" and "Peer certificate cannot be authenticated"), curl doesn't retry the request. At one point between the last exit code `60` and the first exit code `0` it did retry twice

image

File: scripts/docker-helper-4.x/cht-docker-compose.sh [hunk 5]
```diff
@@ -261,23 +262,45 @@ get_system_and_docker_info(){
...
+  if [ "$status" = "60" ]; then
+    echo "false: status is $status"
+  fi
+}
```
Comment (m5r) [inline scripts/docker-helper-4.x/cht-docker-compose.sh:290]:
```suggestion
validate_tls(){
  url=$1
  attempt=$2
  max_retries=3
  # by default curl validates TLS.  If we get back  60 then TLS isn't valid:
  # exitcode: https://everything.curl.dev/usingcurl/verbose/writeout.html
  # 60: https://everything.curl.dev/cmdline/exitcode.html
  status=$(curl --retry 3 --write-out "%{exitcode}" -qs  "$url" -o /dev/null)
  if [ "$status" = "60" ]; then
    if [ $attempt -gt $max_retries ]; then
      echo "false: status is $status"
    else
      duration_seconds=$((1 * 2 ** $attempt))
      sleep $duration_seconds
      validate_tls $url $(($attempt+1))
    fi
  fi
}
```

this ad-hoc exponential backoff retry mechanism works on my machine, let me know what's your take on it

**Review by m5r — CHANGES_REQUESTED** [review]

File: scripts/docker-helper-4.x/cht-docker-compose.sh (hunk 4 above)
Comment (mrjones-plip) [reply scripts/docker-helper-4.x/cht-docker-compose.sh]: ah - great sleuthing! I think we can use your manual retry method below.  thank you!

**Comment by mrjones-plip** [comment]
@m5r - reviewing your fixes in 7945bfe  i saw we were doing `1 * 2 ** $iteration` which seemed a bit verbose because `1 * 2` will always be `2`, so took out `1 *`.  

Also, I fixed some shellcheck [globbing](https://github.com/koalaman/shellcheck/wiki/SC2086) and [arithmetic formatting](https://github.com/koalaman/shellcheck/wiki/SC2004)

Finally - there was a logic error I think?  I think if we don't get back a `0` we want to try again.  If we only try again with a `60` (self signed cert) it won't cover your user case of when you get back `7` (Failed to connect) as the nginx service/container is rebooting.

see cf203e41b  and deeefad78

File: scripts/docker-helper-4.x/cht-docker-compose.sh [hunk 6]
```diff
@@ -468,9 +499,14 @@ while [[ "$running" != "true" ]]; do
   running=$(is_nginx_running "$nginxContainerId")
 done
 
-docker exec "$nginxContainerId" bash -c "curl -s -o /etc/nginx/private/cert.pem https://local-ip.medicmobile.org/fullchain"  2>/dev/null
-docker exec "$nginxContainerId" bash -c "curl -s -o /etc/nginx/private/key.pem https://local-ip.medicmobile.org/key"  2>/dev/null
-docker exec "$nginxContainerId" bash -c "nginx -s reload"  2>/dev/null
+# output a new line with echo and then update certs every time we run to ensure they're current
+echo;update_nginx_local_ip_tls_cert "$nginxContainerId"
+if [ "$(validate_tls "$projectURL")" ];then
```
Comment (m5r) [inline scripts/docker-helper-4.x/cht-docker-compose.sh:504]:
```suggestion
if [ "$(validate_tls "$projectURL" 0)" ];then
```

don't forget the initial value for `$attempt` :eyes:

**Review by m5r — APPROVED** [review]
:+1: feel free to merge after you've committed that `$attempt` variable initialization

File: scripts/docker-helper-4.x/cht-docker-compose.sh (hunk 6 above)
Comment (mrjones-plip) [reply scripts/docker-helper-4.x/cht-docker-compose.sh:504]: yikes - thanks!

**Pull request merged**